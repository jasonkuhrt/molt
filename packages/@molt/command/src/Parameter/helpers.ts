import type { ObjectSet } from '../lib/prelude/prelude.js'
import type { Type } from '../Type/index.js'
import type { Builder, BuilderState } from './constructors/basic.js'
import type { FlagName } from '@molt/types'
import type { Simplify } from 'type-fest'

export interface BuilderBase<S> {
  [BuilderStateSymbol]: S
}

export type Value = string | boolean | number

export const getBuilderState = <S>(builder: BuilderBase<S>): Simplify<GetBuilderState<BuilderBase<S>>> =>
  builder[BuilderStateSymbol]

export type GetBuilderState<B extends BuilderBase<any>> = B[BuilderStateSymbol]

export const BuilderStateSymbol = Symbol(`BuilderState`)

export type BuilderStateSymbol = typeof BuilderStateSymbol

export type BuilderStateMinimum = ObjectSet<BuilderState, 'name', FlagName.Types.FlagNames>

export type Infer<$Builder extends Builder> = _Infer<GetBuilderState<$Builder>>

// prettier-ignore
type _Infer<$BuilderState extends BuilderState> =
  $BuilderState extends { type: infer $Type extends Type.Type }
    ? Type.Infer<$Type>
    : never
