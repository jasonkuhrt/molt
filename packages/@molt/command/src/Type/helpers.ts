import type { Optionality } from '../lib/Pam/parameter.js'
import type { PromptEngine } from '../lib/PromptEngine/PromptEngine.js'
import type { Tex } from '../lib/Tex/index.js'
import type { ValidationResult } from './Type.js'
import type { Effect, Either } from 'effect'

export const TypeSymbol = Symbol(`type`)

export const runtimeIgnore: any = true

export type TypeSymbol = typeof TypeSymbol

export interface Type<T = any> {
  _tag: string
  description: null | string
  [TypeSymbol]: T
  validate: (value: unknown) => ValidationResult<T>
  transform?: (value: T) => T
  help: (settings?: any) => string | Tex.Block
  display: () => string
  // TODO use Either type here
  deserialize: (serializedValue: string) => Either.Either<Error, T>
  prompt: (params: {
    channels: PromptEngine.Channels
    optionality: Optionality
    prompt: string
    marginLeft?: number
  }) => Effect.Effect<never, never, T | undefined>
}

export type Infer<T extends Type<any>> = T[TypeSymbol]
