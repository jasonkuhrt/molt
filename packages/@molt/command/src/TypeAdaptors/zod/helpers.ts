import type {
  SomeBasicType,
  SomeType,
  SomeUnionType,
  SomeUnionTypeScalar,
} from '../../CommandParameter/types.js'
import { ZodHelpers } from '../../lib/zodHelpers/index.js'
import { type ZodNumberCheck, type ZodStringCheck } from '../../lib/zodHelpers/index_.js'
import type { Type } from '../../Type/index.js'
import { Alge } from 'alge'
import { z } from 'zod'

export const analyzeZodType = (zodType: SomeType) => {
  return analyzeZodType_(zodType, zodType.description ?? null)
}

// prettier-ignore
type ZodTypeToType<ZT extends SomeBasicType | SomeUnionType> =
  ZT extends z.ZodOptional<infer T>       ? ZodTypeToType<T> :
  ZT extends z.ZodDefault<infer T>        ? ZodTypeToType<T> :
  ZT extends z.ZodLiteral<any>            ? Type.Scalar.Literal :
  ZT extends z.ZodString                  ? Type.Scalar.String :
  ZT extends z.ZodBoolean                 ? Type.Scalar.Boolean :
  ZT extends z.ZodNumber                  ? Type.Scalar.Number :
  ZT extends z.ZodEnum<any>               ? Type.Scalar.Enumeration :
  ZT extends z.ZodNativeEnum<any>         ? Type.Scalar.Enumeration :
  ZT extends z.ZodUnion<any>              ? Type.Union :
                                            never

const analyzeZodType_ = <ZT extends SomeType>(
  zodType: ZT,
  description: null | string,
): {
  type: ZodTypeToType<ZT>
  // type: ZT
  description: null | string
} => {
  if (zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodDefault) {
    return analyzeZodType_(zodType._def.innerType as ZT, description)
  }

  if (zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodOptional) {
    return analyzeZodType_(zodType._def.innerType as ZT, description)
  }

  // @ts-expect-error todo
  const type: null | Type.Type = ZodHelpers.isUnion(zodType)
    ? analyzeZodUnionType(zodType)
    : ZodHelpers.isLiteral(zodType)
    ? { _tag: `TypeLiteral`, value: zodType._def.value }
    : ZodHelpers.isString(zodType)
    ? { _tag: `TypeString`, ...mapZodStringChecks(zodType._def.checks) }
    : zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodBoolean
    ? { _tag: `TypeBoolean` }
    : ZodHelpers.isNumber(zodType)
    ? { _tag: `TypeNumber`, ...mapZodNumberChecks(zodType._def.checks) }
    : ZodHelpers.isNativeEnum(zodType)
    ? { _tag: `TypeEnum`, members: Object.values(zodType._def.values) }
    : ZodHelpers.isEnum(zodType)
    ? { _tag: `TypeEnum`, members: zodType._def.values }
    : null

  if (!type) throw new Error(`Unsupported zod type: ${zodType._def.typeName}`)

  return {
    description,
    type: type as ZodTypeToType<ZT>,
  }
}

export const analyzeZodUnionType = (zodType: SomeUnionTypeScalar) => {
  return {
    _tag: `TypeUnion` as const,
    members: zodType._def.options.map((_) => {
      return analyzeZodType(_)
      // const typeAnalysis = analyzeZodType(_)
      // return {
      //   // zodType: _,
      //   description: typeAnalysis.description,
      //   type: typeAnalysis.type,
      // }
    }),
  }
}

const mapZodNumberChecks = (checks: z.ZodNumberCheck[]) => {
  return checks
    .map((_): ZodNumberCheck => ({ _tag: _.kind, ..._ }) as any)
    .reduce(
      (acc, check) => {
        return Alge.match(check)
          .int(() => ({ ...acc, int: true }))
          .min((check) => ({
            min: check.value,
          }))
          .max((check) => ({
            max: check.value,
          }))
          .finite(() => ({
            finite: true,
          }))
          .multipleOf((check) => ({
            multipleOf: check.value,
          }))
          .done()
      },
      {} as Omit<Type.Scalar.Number, '_tag'>,
    )
}

const mapZodStringChecks = (checks: z.ZodStringCheck[]): Omit<Type.Scalar.String, '_tag'> => {
  return checks
    .map((_): ZodStringCheck => ({ _tag: _.kind, ..._ }) as any)
    .reduce<Omit<Type.Scalar.String, '_tag'>>(
      (acc, check) => {
        return {
          ...acc,
          ...Alge.match(check)
            .regex((check) => ({
              regex: check.regex,
            }))
            .min((check) => ({
              min: check.value,
            }))
            .max((check) => ({
              max: check.value,
            }))
            .url((check) => ({
              pattern: { type: check._tag },
            }))
            .cuid((check) => ({
              pattern: { type: check._tag },
            }))
            .cuid2(() => ({
              pattern: { type: `cuid2` as const },
            }))
            .uuid((check) => ({
              pattern: { type: check._tag },
            }))
            .emoji((_) => ({
              pattern: { type: _._tag },
            }))
            .ip((_) => ({
              pattern: {
                type: _._tag,
                version: _.version
                  ? Alge.match(_.version)
                      .v4(() => 4 as const)
                      .v6(() => 6 as const)
                      .done()
                  : null,
              },
            }))
            .ulid((_) => ({
              pattern: { type: _._tag },
            }))
            .datetime((check) => ({
              pattern: {
                type: `dateTime` as const,
                offset: check.offset,
                precision: check.precision,
              },
            }))
            .email((check) => ({
              pattern: { type: check._tag },
            }))
            .endsWith((check) => ({
              endsWith: check.value,
            }))
            .startsWith((check) => ({
              startsWith: check.value,
            }))
            .length((check) => ({
              length: check.value,
            }))
            .includes((_) => ({
              includes: _.value,
            }))
            // transformations
            .trim(() => ({
              transformations: {
                ...acc.transformations,
                trim: true,
              },
            }))
            .toLowerCase(() => ({
              transformations: {
                ...acc.transformations,
                toCase: `lower` as const,
              },
            }))
            .toUpperCase(() => ({
              transformations: {
                ...acc.transformations,
                toCase: `upper` as const,
              },
            }))
            .done(),
        }
      },
      {} as Omit<Type.Scalar.String, '_tag'>,
    )
}
