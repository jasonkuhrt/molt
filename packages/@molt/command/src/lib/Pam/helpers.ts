import type { Type } from './kinds.js'

// prettier-ignore
export type TypeToValueMapping<T extends Type> =
  T extends Type.Scalar.Boolean         ? boolean :
  T extends Type.Scalar.Number          ? number  :
  T extends Type.Scalar.String          ? string  :
  T extends Type.Scalar.Enumeration     ? string  :
  T extends Type.Union                  ? boolean | number | string :
                                          never
