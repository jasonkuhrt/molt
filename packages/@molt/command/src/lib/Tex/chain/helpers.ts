import type { Block } from '../nodes/block.js'
import type { List } from '../nodes/list.js'
import type { Node } from '../nodes/node.js'
import type { Table } from '../nodes/table.js'
import type { BlockBuilder } from './block.js'
import type { ListBuilder } from './list.js'
import type { RootBuilder } from './root.js'
import type { TableBuilder } from './table.js'

export type NodeImplementor<B extends Builder> = ($: B) => null | B

export type Builder = RootBuilder | BlockBuilder | TableBuilder | ListBuilder

export interface BuilderInternal<N = Node> {
  _: {
    node: N
  }
}

// prettier-ignore
export const toInternalBuilder = <Builder extends BlockBuilder<null>|TableBuilder|ListBuilder|null>(builder: Builder):
  Builder extends null               ? null                   :
  Builder extends BlockBuilder<null> ? BuilderInternal<Block> :
  Builder extends TableBuilder       ? BuilderInternal<Table> :
  Builder extends ListBuilder        ? BuilderInternal<List>  :
                                       never => 
  builder as any
