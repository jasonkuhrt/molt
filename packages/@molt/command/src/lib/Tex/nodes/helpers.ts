import { Text } from '../../Text/index.js'
import type { BlockParameters } from './block.js'

export interface RenderContext {
  maxWidth?: undefined | number
  height?: undefined | number
  color?: undefined | ((text: string) => string)
  phase?: undefined | 'inner' | 'outer'
  index: {
    total: number
    isLast: boolean
    isFirst: boolean
    position: number
  }
}

export interface Shape {
  intrinsicWidth: number
  intrinsicHeight: number
  desiredWidth: number | null
}

export const applyPadding = (
  value: string,
  padding: Exclude<BlockParameters['padding'], undefined>,
  context: Pick<RenderContext, 'index'>,
) => {
  value =
    padding.topBetween && !context.index.isFirst
      ? Text.chars.newline.repeat(padding.topBetween) + value
      : value
  value = padding.top ? Text.chars.newline.repeat(padding.top) + value : value
  value = padding.left ? Text.indentBlock(value, Text.chars.space.repeat(padding.left)) : value
  value = padding.bottom ? value + Text.chars.newline.repeat(padding.bottom) : value
  value =
    padding.bottomBetween && !context.index.isLast
      ? value + Text.chars.newline.repeat(padding.bottomBetween)
      : value
  value = padding.right
    ? Text.fromLines(Text.toLines(value).map((_) => _ + Text.chars.space.repeat(padding.right!)))
    : value

  return value
}
