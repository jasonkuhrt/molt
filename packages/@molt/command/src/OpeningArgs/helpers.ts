import { BooleanLookup, negateNamePattern, parseEnvironmentVariableBoolean } from '../helpers.js'
import type { ParameterSpec } from '../ParameterSpec/index.js'
import type { Value } from './types.js'
import { Alge } from 'alge'
import camelCase from 'lodash.camelcase'
import { z } from 'zod'

export const stripeDashPrefix = (flagNameInput: string): string => {
  return flagNameInput.replace(/^-+/, ``)
}

export const zodPassthrough = <T>() => z.any().transform((_) => _ as T)

// prettier-ignore
export const parseRawInput = (name: string, rawValue: string, spec: ParameterSpec.Output): Value => {
  const parsedValue = parseRawValue(rawValue, spec)
  if (parsedValue === null) {
    const expectedTypes = Alge.match(spec).Union((spec) => spec.types.map(_=>_.type._tag).join(` | `)).else(spec => spec.type._tag)
    throw new Error(`Failed to parse input ${name} with value ${rawValue}. Expected type of ${expectedTypes}.`)
  }
  if (typeof parsedValue === `string`) return { _tag: `string`, value: parsedValue }
  if (typeof parsedValue === `number`) return { _tag: `number`, value: parsedValue }
  if (typeof parsedValue === `boolean`){
  // dump(isEnvarNegated(name, spec))
  return { _tag: `boolean`, value: parsedValue, negated: isEnvarNegated(name, spec) }
  }
  return casesHandled(parsedValue)
}

const casesHandled = (value: never): never => {
  throw new Error(`Unhandled case ${String(value)}`)
}

/**
 * Is the environment variable input negated? Unlike line input the environment can be
 * namespaced so a bit more work is needed to parse out the name pattern.
 */
export const isEnvarNegated = (name: string, spec: ParameterSpec.Output): boolean => {
  const nameWithNamespaceStripped = stripeNamespace(name, spec)
  // dump({ nameWithNamespaceStripped })
  return negateNamePattern.test(nameWithNamespaceStripped)
}

export const isNegated = (name: string): boolean => {
  return negateNamePattern.test(name)
}

const stripeNamespace = (name: string, spec: ParameterSpec.Output): string => {
  for (const namespace of spec.environment?.namespaces ?? []) {
    if (name.startsWith(namespace)) return camelCase(name.slice(namespace.length))
  }
  return name
}

export const parseRawValue = (
  value: string,
  spec: ParameterSpec.Output,
): null | boolean | number | string => {
  return Alge.match(spec)
    .Union((spec) => {
      /**
       * For a union we infer the value to be the type of the first variant type that matches.
       * This means that variant order matters since there are sub/super type relationships.
       * For example a number is a subset of string type. If there is a string and number variant
       * we should first check if the value could be a number, than a string.
       */
      const variantOrder: ParameterSpec.Type['_tag'][] = [
        `TypeNumber`,
        `TypeBoolean`,
        `TypeString`,
        `TypeEnum`,
      ]
      const types = spec.types.sort(
        (a, b) => variantOrder.indexOf(a.type._tag) - variantOrder.indexOf(b.type._tag),
      )
      return (
        types
          .map((_) =>
            Alge.match(_.type)
              .TypeLiteral((_) => parseLiteral(_, value))
              .TypeString(() => value)
              .TypeEnum((_) => parseEnum(_, value))
              .TypeBoolean(() => parseEnvironmentVariableBoolean(value))
              .TypeNumber(() => {
                const result = Number(value)
                return isNaN(result) ? null : result
              })
              .done(),
          )
          .find((parsedValue) => parsedValue !== null) ?? null
      )
    })
    .else((spec) =>
      Alge.match(spec.type)
        .TypeLiteral((_) => parseLiteral(_, value))
        .TypeString(() => value)
        .TypeEnum((_) => parseEnum(_, value))
        .TypeBoolean(() => parseEnvironmentVariableBoolean(value))
        .TypeNumber(() => Number(value))
        .done(),
    )
}

/**
 * Enums can be given a base type of numbers or strings. Examples:
 *
 * - number: `z.nativeEnum(\{ a: 1, b: 2\})`
 * - string: `z.enum(['a','b','c'])`
 *
 * It is not possible to have an enum that mixes numbers and strings.
 *
 * When we receive a raw value, we infer its base  type based on checking the type first member of the enum.
 */
export const parseEnum = (spec: ParameterSpec.TypeEnum, value: string): string | number => {
  const isNumberEnum = spec.members.find((_) => typeof _ === `number`)
  if (isNumberEnum) return Number(value)
  return value
}

export const parseLiteral = (spec: ParameterSpec.TypeLiteral, value: string): boolean | string | number => {
  if (typeof spec.value === `string`) return value
  if (typeof spec.value === `number`) return Number(value)
  if (typeof spec.value === `boolean`) {
    const v = (BooleanLookup as Record<string, boolean>)[value]
    if (!v) throw new Error(`Invalid boolean literal value: ${value}`)
    return v
  }
  return casesHandled(spec.value)
}
