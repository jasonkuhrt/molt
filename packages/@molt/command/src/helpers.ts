import camelCase from 'lodash.camelcase'
import { z } from 'zod'

export const BooleanLookup = {
  true: true,
  false: false,
} as const

export const environmentVariableBooleanLookup = {
  ...BooleanLookup,
  '1': true,
  '0': false,
} as const

export const stripeDashPrefix = (flagNameInput: string): string => {
  return flagNameInput.replace(/^-+/, ``)
}

export const zodPassthrough = <T>() => z.any().transform((_) => _ as T)

export type Values<T> = T[keyof T]

export const getLowerCaseEnvironment = (): NodeJS.ProcessEnv => lowerCaseObjectKeys(process.env)

export const lowerCaseObjectKeys = (obj: object) =>
  Object.fromEntries(Object.entries(obj).map(([k, v]) => [k.toLowerCase(), v]))

export const parseEnvironmentVariableBoolean = (value: string): boolean | null =>
  // @ts-expect-error ignore
  // eslint-disable-next-line
  environmentVariableBooleanLookup[value] ?? null

export const parseEnvironmentVariableBooleanOrThrow = (value: string) => {
  const result = parseEnvironmentVariableBoolean(value)
  if (result === null) {
    throw new Error(`Invalid boolean value: ${value}`)
  }
  return result
}

export const negateNamePattern = /^no([A-Z].+)/

export const stripeNegatePrefix = (name: string): null | string => {
  // eslint-disable-next-line
  const withoutPrefix = name.match(negateNamePattern)?.[1]!
  if (!withoutPrefix) return null
  const withCamelCase = camelCase(withoutPrefix)
  return withCamelCase
}

export const stripeNegatePrefixLoose = (name: string): string => {
  const result = stripeNegatePrefix(name)
  return result ? result : name
}

export const invertTable = <T>(rows: T[][]): T[][] => {
  const columns: T[][] = []
  for (const row of rows) {
    let i = 0
    for (const col of row) {
      const column = columns[i] || []
      column.push(col)
      columns[i] = column
      i++
    }
  }
  return columns
}

export const entries = <O extends Record<string, unknown>>(
  obj: O,
): Exclude<{ [k in keyof O]: [k, O[k]] }[keyof O], undefined>[] => Object.entries(obj) as any

export const casesExhausted = (_: never): never => {
  throw new Error(`Cases exhausted: ${_}`) // eslint-disable-line
}
