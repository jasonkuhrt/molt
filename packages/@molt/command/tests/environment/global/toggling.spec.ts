import { expect } from 'vitest'
import { it } from 'vitest'
import { $, s } from '../../_/helpers.js'
import { createState, environmentManager } from '../__helpers__.js'

const output = createState<string>({
  value: (values) => values.join(``),
})

it(`is enabled by default`, () => {
  environmentManager.set(`cli_parameter_foo`, `bar`)
  const args = $.parameter(`--foo`, s).parse({ line: [] })
  expect(args).toMatchObject({ foo: `bar` })
})
it(`can be enabled by settings`, () => {
  environmentManager.set(`cli_param_foo`, `bar`)
  const args1 = $.parameter(`--foo`, s)
    .settings({ parameters: { environment: true } })
    .parse({ line: [] })
  expect(args1).toMatchObject({ foo: `bar` })
  const args2 = $.parameter(`--foo`, s)
    .settings({ parameters: { environment: { $default: true } } })
    .parse({ line: [] })
  expect(args2).toMatchObject({ foo: `bar` })
  const args3 = $.parameter(`--foo`, s)
    .settings({ parameters: { environment: { $default: { enabled: true } } } })
    .parse({ line: [] })
  expect(args3).toMatchObject({ foo: `bar` })
})
it(`can be enabled by environment`, () => {
  environmentManager.set(`ClI_settings_READ_arguments_FROM_ENVIRONMENT`, `true`)
  environmentManager.set(`cli_param_foo`, `bar`)
  const args = $.parameter(`--foo`, s).parse({ line: [] })
  expect(args).toMatchObject({ foo: `bar` })
})
it(`can be disabled by environment`, () => {
  environmentManager.set(`ClI_settings_READ_arguments_FROM_ENVIRONMENT`, `false`)
  environmentManager.set(`cli_param_foo`, `foo_env`)
  const args = $.parameter(`--foo`, s.default(`foo_default`)).parse({ line: [] })
  expect(args).toMatchObject({ foo: `foo_default` })
})
it(`environment supersedes settings`, () => {
  expect(() =>
    $.parameter(`--foo`, s)
      .settings({
        onOutput: output.set,
        parameters: { environment: true },
        helpOnNoArguments: false,
        onError: `throw`,
      })
      .parse({
        line: [],
        environment: { ClI_settings_READ_arguments_FROM_ENVIRONMENT: `false`, cli_param_foo: `bar` },
      })
  ).toThrowErrorMatchingSnapshot()
  expect(output.value).toMatchSnapshot()
})
