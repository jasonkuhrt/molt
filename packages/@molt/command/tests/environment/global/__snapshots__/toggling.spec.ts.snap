// Vitest Snapshot v1

exports[`environment supersedes settings 1`] = `
[
  [
    "Cannot run command, you made some mistakes:

Missing argument for flag \\"foo\\".

Here are the docs for this command:

",
  ],
  [
    "
PARAMETERS

  [4m[90mName[39m[24m   [4m[90mType[39m[24m         [4m[90mDefault[39m[24m      [4m[90mEnvironment (1)[39m[24m                         
  
  [32mfoo[39m    [32mstring[39m       [1m[30m[101m REQUIRED [49m[39m[22m   [34m✓[39m          
  
[90m  NOTES[39m
[90m  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━[39m
[90m  (1) Parameters can be passed arguments via environment variables. Command line a             [39m
[90m      rguments take precedence. Environment variable names are snake cased version             [39m
[90m      s of the parameter name (or its aliases), case insensitive. They must be pre             [39m
[90m      fixed with [34mCLI_PARAMETER_[39m[90m or [34mCLI_PARAM_[39m[90m (case insensitive). Examples:                    [39m
[90m        → [34mCLI_PARAMETER_[39m[90m[32mFOO[39m[90m=\\"...\\".                                                             [39m
",
  ],
]
`;
